form/data_transfer_objects.rst

```diff
@@ -0,0 +1,347 @@
+.. index::
+   single: Form; Data Transfer Objects
+
+How to use Data Transfer Objects (DTOs)
+=======================================
+
+Data Transfer Objects can be used by forms to separate entities from the
+validation logic of forms.
+Entities should always have a valid state.
+When entities are used as data classes for a form, the data is injected into
+the entity and validated.
+When the validation fails, the invalid data is still left in the entity.
+This can lead to invalid data being saved in the database.
```

Comment:
Or maybe unexpected errors? if data is invalid it probably won't even get to be saved in the database.

Reply:
It can easily end up in the database.
Example: Email validation fails, entity is saved in a listener or somewhere further down, database field expects a string, so it won't fail on the insert.

Reply:
Good point, but it can also lead to exceptions. So maybe just add `or unexpected exceptions`?